coderd/schedule/autostop.go

```diff
@@ -104,9 +117,30 @@ func CalculateAutostop(ctx context.Context, params CalculateAutostopParams) (Aut
 	if !templateSchedule.UserAutostopEnabled {
 		// The user is not permitted to set their own TTL, so use the template
 		// default.
-		autostop.Deadline = time.Time{}
+		ttl = 0
 		if templateSchedule.DefaultTTL > 0 {
-			autostop.Deadline = now.Add(templateSchedule.DefaultTTL)
+			ttl = templateSchedule.DefaultTTL
+		}
+	}
+
+	if ttl > 0 {
+		// Only apply non-zero TTLs.
+		autostop.Deadline = now.Add(ttl)
+		if params.WorkspaceAutostart != "" {
+			// If the deadline passes the next autostart, we need to extend the deadline to
+			// autostart + deadline. ActivityBumpWorkspace already covers this case
+			// when extending the deadline.
+			//
+			// Situation this is solving.
+			// 1. User has workspace with auto-start at 9:00am, 12 hour auto-stop.
+			// 2. Coder stops workspace at 9pm
+			// 3. User starts workspace at 9:45pm.
+			//	- The initial deadline is calculated to be 9:45am
+			//	- This crosses the autostart deadline, so the deadline is extended to 9pm
+			nextAutostart, ok := NextAutostartSchedule(params.Now, params.WorkspaceAutostart, templateSchedule)
+			if ok && autostop.Deadline.After(nextAutostart) {
+				autostop.Deadline = nextAutostart.Add(ttl)
```

Comment:
Do we need to handle the `min(deadline, max_deadline)` here too?

Reply:
No, that is handled down below.

https://github.com/coder/coder/blob/94872aa9a2b513f9b296ce4a95b8338e85091a0d/coderd/schedule/autostop.go#L238-L238

I'd rather not duplicate that logic since it also handles the case where MaxDeadline is not set.